text/0000-stl-types.md

```diff
@@ -0,0 +1,168 @@
+# Summary
+
+Lots of libraries use allocating standard library types in their API and we need a method for dealing with:
+- `std::vector`
+- `std::string`
+
+The simplest way of dealing with these is to just copy the types entirely to C equivalents in the wrapper functions. This is the method currently proposed for handling strings. For example:
+```c++
+int OIIO_ImageInput_geterror(const OIIO_ImageInput* self, char* _result_buffer_ptr, int _result_buffer_len) {
+    const std::string result = to_cpp(self)->geterror();
+    safe_strcpy(_result_buffer_ptr, result, _result_buffer_len);
+    return result.size();
+}
+```
+In this model, the caller allocates the memory and passes it in to the wrapper. The wrapper then copies from the string returned from C++ to the provided buffer, making sure not to exceed the user-provided length.
+
+This works fine for short arrays, such as error messages where we expect to be able to stack-allocate the buffer and adjust the program if we find it's not large enough, or allocate a reasonably large one once and reuse it throughout the program, but strings and vectors could really be any size. We need a different model.
+
+The crux of the issue is that `vector` and `string` allocate and manage their memory using RAII and provide no facility for releasing ownership of that memory.
+
+In the examples below we'll just use `std::vector`. There are three cases we need to consider:
+
+```c++
+namespace foo {
+std::vector<float> returns();
+void as_mut_param(std::vector<float>& vec);
+void as_const_param(const std::vector<float>& vec);
+}
+```
+
+# Possible solutions
+
+## 1. Two-phase copy
+
+This is a modifcation of the error-message method above. For returning, we make two functions, say `len` and `get`. `len` is the function that actually gets the vector, then stashes it in a thread_local and returns its size, giving the caller an opportunity to allocate memory of the required size which it can pass to `get`, which will copy from the stashed vector:
+
+```c++
+int foo_returns_len() {
+    thread_local _stash = foo::returns();
+    return _stash.size();
+}
+
+void foo_returns_get(float* buffer, int len) {
+    memcpy(buffer, _stash.data(), sizeof(float) * min(_stash.size(), len));
+    _stash = {};
+}
+```
+
+`const_param` just copies into a temporary vector:
+```c++
+void foo_as_const_param(float* buffer, int len) {
+    std::vector tmp{buffer, len};
+    foo::as_const_param(tmp);
+}
+```
+
+`mut_param` is trickier since we potentially need to both pass data and return it. Easiest is probably just not to support passing data at all, in which case it reduces to the same as `returns`.
+
+### Pros
+- It's guaranteed to work and is safe
+- Caller is responsible for all allocations
+
```

Comment:
do you mean the caller is responsible for inducing all allocations? That's how I read the proposal. I don't read that the caller is responsible for allocations ~ which wouldn't make sense, given that the vector and string can cause allocations themselves independent of what the caller intended. 

All of these proposals need to consider allocation boundaries as well. On Windows for example, given Foo.dll and Bar.dll, each bound to unknown standard libraries, we can guarantee that Foo can freely allocate and free a std::string, and that Bar can access the bytes of the string, in a single threaded context. A string allocated by Foo, and deallocated by Bar drops into the realm of undefined behavior, and more typically simply corrupts Foo's heap, and Bar's heap by cross linking freelists.

Reply:
Yes absolutely, this is one of the main things we're trying to work around here. In the first option yeah, vector<string> absolutely will not work.